File: src/main/java/uk/edu/populationfitness/models/genes/valley/ZakharoyGenes.java
package uk.edu.populationfitness.models.genes.valley;

import uk.edu.populationfitness.models.Config;
import uk.edu.populationfitness.models.fastmaths.ExpensiveCalculatedValues;
import uk.edu.populationfitness.models.fastmaths.FastMaths;
import uk.edu.populationfitness.models.fastmaths.ValueCalculator;
import uk.edu.populationfitness.models.genes.bitset.NormalizingBitSetGenes;

public class ZakharoyGenes extends NormalizingBitSetGenes {
    private static class NormalizationRatioCalculator implements ValueCalculator<Double>{
        @Override
        public Double calculateValue(long n) {
            double value = (n * (n + 1.0) ) / 2.0;
            return 100 * n + 25 * (value * value + 35.0 * FastMaths.pow(value, 4));
        }
    }

    private static final ExpensiveCalculatedValues<Double> NormalizationRatios = new ExpensiveCalculatedValues(new NormalizationRatioCalculator());

    public ZakharoyGenes(Config config) {
        super(config, 10.0);
    }

    @Override
    protected double calculateNormalizationRatio(int n) {
        return NormalizationRatios.findOrCalculate(n);
    }

    @Override
    protected double calculateFitnessFromIntegers(long[] integer_values) {
        /*
          f left (x right ) = sum from {i=1} to {n} {{x} rsub {i} rsup {2}} + {left (sum from {i=1} to {n} {0.5 i  {x} rsub {i}} right )} ^ {2} + {left (sum from {i=1} to {n} {0.5 i  {x} rsub {i}} right )} ^ {4}
         */
        double fitness = 0;

        double sum = 0;

        for(int i = 1; i <= integer_values.length; i++){
            double x = interpolate(integer_values[i - 1]);
            fitness += x * x;
            sum += 0.5 * i * x;
        }

        fitness += (sum * sum) + (sum * sum * sum * sum);

        return fitness;
    }
}

File: DotNet/PopulationFitness/PopulationFitness/Models/Genes/Valley/ZakharoyGenes.cs
using PopulationFitness.Models.FastMaths;
using PopulationFitness.Models.Genes.BitSet;

namespace PopulationFitness.Models.Genes.Valley
{
    public class ZakharoyGenes : NormalizingBitSetGenes
    {
        private class NormalizationRatioCalculator : IValueCalculator<double>
        {
            public double CalculateValue(long n)
            {
                double value = (n * (n + 1.0)) / 2.0;
                return 100 * n + 25 * (value * value + 35.0 * FastMaths.FastMaths.Pow(value, 3));
            }
        }

        private static readonly ExpensiveCalculatedValues<double> NormalizationRatios = new ExpensiveCalculatedValues<double>(new NormalizationRatioCalculator());

        public ZakharoyGenes(Config config) : base(config, 10.0)
        { }

        protected override double CalculateNormalizationRatio(int n)
        {
            return NormalizationRatios.FindOrCalculate(n);
        }

        protected override double CalculateFitnessFromIntegers(long[] integer_values)
        {
            /*
              f left (x right ) = sum from {i=1} to {n} {{x} rsub {i} rsup {2}} + {left (sum from {i=1} to {n} {0.5 i  {x} rsub {i}} right )} ^ {2} + {left (sum from {i=1} to {n} {0.5 i  {x} rsub {i}} right )} ^ {4}
             */
            double fitness = 0;

            double sum = 0;

            for (int i = 1; i <= integer_values.Length; i++)
            {
                double x = Interpolate(integer_values[i - 1]);
                fitness += x * x;
                sum += 0.5 * i * x;
            }

            fitness += (sum * sum) + (sum * sum * sum * sum);

            return fitness;
        }
    }
}
